File: src/main/java/no/nav/sosialhjelp/soknad/consumer/common/rest/RestUtils.java
package no.nav.sosialhjelp.soknad.consumer.common.rest;

/* Originally from common-java-modules (no.nav.sbl.dialogarena.common.rest) */

import lombok.SneakyThrows;
import no.nav.sosialhjelp.soknad.consumer.common.json.JsonProvider;
import no.nav.sosialhjelp.soknad.consumer.common.rest.client.MetricsConnectorProvider;
import org.glassfish.jersey.client.ClientConfig;
import org.glassfish.jersey.client.JerseyClientBuilder;

import javax.net.ssl.SSLContext;
import javax.ws.rs.client.Client;

import static org.glassfish.jersey.client.ClientProperties.CONNECT_TIMEOUT;
import static org.glassfish.jersey.client.ClientProperties.FOLLOW_REDIRECTS;
import static org.glassfish.jersey.client.ClientProperties.READ_TIMEOUT;

public final class RestUtils {

    public static final String CSRF_COOKIE_NAVN = "NAV_CSRF_PROTECTION";
    public static final RestConfig DEFAULT_CONFIG = new RestConfig.Builder().build();

    private RestUtils() {
    }

    private static ClientConfig createClientConfig(RestConfig restConfig, String metricName) {
        ClientLogFilter clientLogFilter = new ClientLogFilter(ClientLogFilter.ClientLogFilterConfig.builder()
                .disableMetrics(restConfig.getDisableMetrics())
                .disableParameterLogging(restConfig.getDisableParameterLogging())
                .metricName(metricName)
                .build()
        );

        ClientConfig clientConfig = new ClientConfig();
        clientConfig.register(new JsonProvider());
        clientConfig.register(clientLogFilter);
        clientConfig.property(FOLLOW_REDIRECTS, false);
        clientConfig.property(CONNECT_TIMEOUT, restConfig.getConnectTimeout());
        clientConfig.property(READ_TIMEOUT, restConfig.getReadTimeout());
        clientConfig.connectorProvider(new MetricsConnectorProvider(clientConfig.getConnectorProvider(), clientLogFilter));
        return clientConfig;
    }

    public static Client createClient() {
        return createClient(DEFAULT_CONFIG, getMetricName());
    }

    public static Client createClient(RestConfig restConfig) {
        return createClient(restConfig, getMetricName());
    }

    private static Client createClient(RestConfig restConfig, String metricName) {
        return new JerseyClientBuilder()
                .sslContext(defaultSSLContext())
                .withConfig(createClientConfig(restConfig, metricName))
                .build();
    }

    @SneakyThrows
    private static SSLContext defaultSSLContext() {
        return SSLContext.getDefault();
    }

    private static String getMetricName() {
        StackTraceElement element = Thread.currentThread().getStackTrace()[3];
        return String.format("rest.client.%s.%s", element.getClassName(), element.getMethodName());
    }
}

File: src/main/kotlin/no/nav/sosialhjelp/soknad/common/rest/RestUtils.kt
package no.nav.sosialhjelp.soknad.common.rest

/* Originally from common-java-modules (no.nav.sbl.dialogarena.common.rest) */

import no.nav.sosialhjelp.soknad.common.json.JsonProvider
import no.nav.sosialhjelp.soknad.common.rest.ClientLogFilter.ClientLogFilterConfig
import org.glassfish.jersey.client.ClientConfig
import org.glassfish.jersey.client.ClientProperties
import org.glassfish.jersey.client.JerseyClientBuilder
import javax.net.ssl.SSLContext
import javax.ws.rs.client.Client

object RestUtils {
    const val CSRF_COOKIE_NAVN = "NAV_CSRF_PROTECTION"
    private val DEFAULT_CONFIG = RestConfig()

    private fun createClientConfig(restConfig: RestConfig, metricName: String): ClientConfig {
        val clientLogFilter = ClientLogFilter(
            ClientLogFilterConfig(
                metricName = metricName,
                disableMetrics = restConfig.disableMetrics,
                disableParameterLogging = restConfig.disableParameterLogging
            )
        )
        val clientConfig = ClientConfig()
        clientConfig.register(JsonProvider())
        clientConfig.register(clientLogFilter)
        clientConfig.property(ClientProperties.FOLLOW_REDIRECTS, false)
        clientConfig.property(ClientProperties.CONNECT_TIMEOUT, restConfig.connectTimeout)
        clientConfig.property(ClientProperties.READ_TIMEOUT, restConfig.readTimeout)
        clientConfig.connectorProvider(MetricsConnectorProvider(clientConfig.connectorProvider, clientLogFilter))
        return clientConfig
    }

    fun createClient(): Client {
        return createClient(DEFAULT_CONFIG, metricName)
    }

    fun createClient(restConfig: RestConfig): Client {
        return createClient(restConfig, metricName)
    }

    private fun createClient(restConfig: RestConfig, metricName: String): Client {
        return JerseyClientBuilder()
            .sslContext(defaultSSLContext())
            .withConfig(createClientConfig(restConfig, metricName))
            .build()
    }

    private fun defaultSSLContext(): SSLContext {
        return SSLContext.getDefault()
    }

    private val metricName: String
        get() {
            val element = Thread.currentThread().stackTrace[3]
            return String.format("rest.client.%s.%s", element.className, element.methodName)
        }
}
